polkadot/node/core/runtime-api/src/lib.rs

```diff
@@ -339,9 +357,18 @@ where
 			None => return,
 		};
 
+		// Every-time we fetch a new session info from the runtime, let's check the node features as
+		// well to make sure things are compatible.
+		let should_check_node_features = match request {
+			Request::SessionInfo(session_index, _) =>
+				Some((relay_parent, session_index, self.client.clone(), self.metrics.clone())),
+			_ => None,
+		};
+
 		let request = async move {
 			let result = make_runtime_api_request(client, metrics, relay_parent, request).await;
 			let _ = sender.send(result);
+			Self::maybe_check_node_features(should_check_node_features).await;
```

Comment:
I don't think we should do that in the runtime-api subsystem itself. This logic can be extracted into a function in subsystem-util, which can be a wrapper around `request_node_features` and that can do some additional error handling, e.g. returning `false` on `Error::NotSupported`, etc.
Cf. #858

Reply:
that already exists. see: `request_node_features`.

What this PR does is to just add a check to see if a new feature is enabled in the runtime that the current node version does not support (to log an warning so that validators know to update).

Reply:
Ok, I thought `request_node_features` was generated by a macro. In that case, it should be part of that function.
The argument here is about separation of concerns. runtime-api subsystem should be a low-level abstraction and doesn't need to know about internals of the types. Let's not so a premature optimization here. It can be done on a higher level up.

Reply:
I think moving the features in `subsystem-util::runtime::ExtendedSessionInfo` is a step in that direction. I thinkg that's a good sport to implement this check.

Reply:
Great idea with moving into `ExtendedSessionInfo` , I think it looks  better now.